massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,659 @@
+    ) -> (Slot, KeyPair, SecuredHeader, SecuredHeader, SecuredHeader) {
+        let keypair = KeyPair::generate();
+
+        let slot = Slot::new(2, 1);
+        let parents_1: Vec<BlockId> = (0..THREAD_COUNT)
+            .map(|i| BlockId(Hash::compute_from(&[i])))
+            .collect();
+        let parents_2: Vec<BlockId> = (0..THREAD_COUNT)
+            .map(|i| BlockId(Hash::compute_from(&[i + 1])))
+            .collect();
+        let parents_3: Vec<BlockId> = (0..THREAD_COUNT)
+            .map(|i| BlockId(Hash::compute_from(&[i + 2])))
+            .collect();
+
+        let endorsement_1 = Endorsement {
+            slot: Slot::new(1, 1),
+            index: 1,
+            endorsed_block: BlockId(Hash::compute_from("blk1".as_bytes())),
+        };
+        let s_endorsement_1 =
+            Endorsement::new_verifiable(endorsement_1, EndorsementSerializerLW::new(), &keypair)
+                .unwrap();
+
+        let block_header_1 = BlockHeader {
+            slot,
+            parents: parents_1,
+            operation_merkle_root: Hash::compute_from("mno".as_bytes()),
+            endorsements: vec![s_endorsement_1.clone()],
+        };
+
+        // create header
+        let s_block_header_1 = BlockHeader::new_verifiable::<BlockHeaderSerializer, BlockId>(
+            block_header_1,
+            BlockHeaderSerializer::new(),
+            &keypair,
+        )
+        .expect("error while producing block header");
+
+        let block_header_2 = BlockHeader {
+            slot,
+            parents: parents_2,
+            operation_merkle_root: Hash::compute_from("mno".as_bytes()),
+            endorsements: vec![s_endorsement_1.clone()],
+        };
+
+        // create header
+        let s_block_header_2 = BlockHeader::new_verifiable::<BlockHeaderSerializer, BlockId>(
+            block_header_2,
+            BlockHeaderSerializer::new(),
+            &keypair,
+        )
+        .expect("error while producing block header");
+
+        let block_header_3 = BlockHeader {
+            slot,
+            parents: parents_3,
+            operation_merkle_root: Hash::compute_from("mno".as_bytes()),
+            endorsements: vec![s_endorsement_1.clone()],
+        };
+
+        // create header
+        let s_block_header_3 = BlockHeader::new_verifiable::<BlockHeaderSerializer, BlockId>(
+            block_header_3,
+            BlockHeaderSerializer::new(),
+            &keypair,
+        )
+        .expect("error while producing block header");
+
+        return (
+            slot,
+            keypair,
+            s_block_header_1.clone(),
+            s_block_header_2,
+            s_block_header_3,
+        );
+    }
+
+    #[test]
+    fn test_block_header_denunciation() {
+        // Create an block header denunciation and check if it is valid
+        let (_slot, _keypair, s_block_header_1, s_block_header_2, s_block_header_3) =
+            gen_block_headers_for_denunciation();
```

Comment:
You may want to add (and extend?) something like this:

```suggestion
    #[test]
    fn test_block_header_denunciation_invalid() {
        // Create a block header denunciation
        let (_slot, _keypair, s_block_header_1, _s_block_header_2, _s_block_header_3) =
            gen_block_headers_for_denunciation();

        let denunciation = Denunciation::try_from((&s_block_header_1, &s_block_header_1));
        assert!(matches!(denunciation, Err(DenunciationError::InvalidInput)));
    }

```

Reply:
done